// zkverifyjs/dist/esm/proofTypes/groth16/formatter/arkworks/index.js
import { isGroth16Config } from "../../../../utils/helpers/index.js";
import { extractCurve } from "../utils.js";
/**
 * Returns proof as-is, just extracts fields.
 */
export const formatProof = (proof, options) => {
  if (!isGroth16Config(options)) {
    throw new Error('Expected Groth16 config but received invalid configuration.');
  }
  if (!proof || typeof proof !== 'object' || !('proof' in proof) || !proof.proof || typeof proof.proof.a !== 'string' || typeof proof.proof.b !== 'string' || typeof proof.proof.c !== 'string' || typeof proof.curve !== 'string') {
    const snippet = JSON.stringify(proof).slice(0, 80);
    throw new Error(`Invalid Arkworks proof format. Expected { curve, proof: { a, b, c } }. Snippet: "${snippet}..."`);
  }
  const ark = proof;
  const curve = extractCurve(options.config.curve);
  return {
    curve,
    proof: {
      a: ark.proof.a,
      b: ark.proof.b,
      c: ark.proof.c
    }
  };
};
/**
 * Returns VK as-is, trusting Arkworks formatting is already compatible.
 */
export const formatVk = (vk, options) => {
  if (!isGroth16Config(options)) {
    throw new Error('Expected Groth16 config but received invalid configuration.');
  }
  const curve = extractCurve(options.config.curve);
  const result = {
    ...vk,
    curve
  };
  return result;
};
/**
 * Arkworks public signals require no transformation.
 */
export const formatPubs = pubs => {
  return pubs;
};
export default {
  formatProof,
  formatVk,
  formatPubs
};
