// zkverifyjs/dist/commonjs/proofTypes/groth16/formatter/arkworks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatPubs = exports.formatVk = exports.formatProof = void 0;
const helpers_1 = require("../../../../utils/helpers");
const utils_1 = require("../utils");
/**
 * Returns proof as-is, just extracts fields.
 */
const formatProof = (proof, options) => {
    if (!(0, helpers_1.isGroth16Config)(options)) {
        throw new Error('Expected Groth16 config but received invalid configuration.');
    }
    if (!proof ||
        typeof proof !== 'object' ||
        !('proof' in proof) ||
        !proof.proof ||
        typeof proof.proof.a !== 'string' ||
        typeof proof.proof.b !== 'string' ||
        typeof proof.proof.c !== 'string' ||
        typeof proof.curve !== 'string') {
        const snippet = JSON.stringify(proof).slice(0, 80);
        throw new Error(`Invalid Arkworks proof format. Expected { curve, proof: { a, b, c } }. Snippet: "${snippet}..."`);
    }
    const ark = proof;
    const curve = (0, utils_1.extractCurve)(options.config.curve);
    return {
        curve,
        proof: {
            a: ark.proof.a,
            b: ark.proof.b,
            c: ark.proof.c,
        },
    };
};
exports.formatProof = formatProof;
/**
 * Returns VK as-is, trusting Arkworks formatting is already compatible.
 */
const formatVk = (vk, options) => {
    if (!(0, helpers_1.isGroth16Config)(options)) {
        throw new Error('Expected Groth16 config but received invalid configuration.');
    }
    const curve = (0, utils_1.extractCurve)(options.config.curve);
    const result = {
        ...vk,
        curve,
    };
    return result;
};
exports.formatVk = formatVk;
/**
 * Arkworks public signals require no transformation.
 */
const formatPubs = (pubs) => {
    return pubs;
};
exports.formatPubs = formatPubs;
exports.default = {
    formatProof: exports.formatProof,
    formatVk: exports.formatVk,
    formatPubs: exports.formatPubs,
};
