// zkverifyjs/dist/commonjs/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PUBLIC_ZK_VERIFY_EVENTS = exports.RuntimeVersion = exports.Destination = exports.ProofSecurityRules = exports.AggregateSecurityRules = exports.CurveType = exports.Library = exports.Plonky2HashFunction = exports.UltrahonkVariant = exports.Risc0Version = exports.ZkVerifyEvents = exports.TransactionStatus = exports.TransactionType = void 0;
var TransactionType;
(function (TransactionType) {
    TransactionType[TransactionType["Verify"] = 1] = "Verify";
    TransactionType[TransactionType["VKRegistration"] = 2] = "VKRegistration";
    TransactionType[TransactionType["DomainRegistration"] = 3] = "DomainRegistration";
    TransactionType[TransactionType["DomainHold"] = 4] = "DomainHold";
    TransactionType[TransactionType["DomainUnregister"] = 5] = "DomainUnregister";
    TransactionType[TransactionType["Aggregate"] = 6] = "Aggregate";
    TransactionType[TransactionType["BatchVerify"] = 7] = "BatchVerify";
    TransactionType[TransactionType["DomainAddSubmitters"] = 8] = "DomainAddSubmitters";
    TransactionType[TransactionType["DomainRemoveSubmitters"] = 9] = "DomainRemoveSubmitters";
})(TransactionType || (exports.TransactionType = TransactionType = {}));
var TransactionStatus;
(function (TransactionStatus) {
    TransactionStatus["Broadcast"] = "broadcast";
    TransactionStatus["Dropped"] = "dropped";
    TransactionStatus["Error"] = "error";
    TransactionStatus["Finalized"] = "finalized";
    TransactionStatus["InBlock"] = "inBlock";
    TransactionStatus["Invalid"] = "invalid";
    TransactionStatus["Pending"] = "pending";
    TransactionStatus["Retracted"] = "retracted";
    TransactionStatus["Usurped"] = "usurped";
})(TransactionStatus || (exports.TransactionStatus = TransactionStatus = {}));
var ZkVerifyEvents;
(function (ZkVerifyEvents) {
    ZkVerifyEvents["AggregationComplete"] = "aggregationComplete";
    ZkVerifyEvents["NewAggregationReceipt"] = "newAggregationReceipt";
    ZkVerifyEvents["Broadcast"] = "broadcast";
    ZkVerifyEvents["CannotAggregate"] = "cannotAggregate";
    ZkVerifyEvents["DomainStateChanged"] = "domainStateChanged";
    ZkVerifyEvents["ErrorEvent"] = "error";
    ZkVerifyEvents["Finalized"] = "finalized";
    ZkVerifyEvents["IncludedInBlock"] = "includedInBlock";
    ZkVerifyEvents["NewDomain"] = "newDomain";
    ZkVerifyEvents["NewProof"] = "newProof";
    ZkVerifyEvents["ProofVerified"] = "proofVerified";
    ZkVerifyEvents["Unsubscribe"] = "unsubscribe";
    ZkVerifyEvents["VkRegistered"] = "vkRegistered";
})(ZkVerifyEvents || (exports.ZkVerifyEvents = ZkVerifyEvents = {}));
var Risc0Version;
(function (Risc0Version) {
    Risc0Version["V2_1"] = "V2_1";
    Risc0Version["V2_2"] = "V2_2";
    Risc0Version["V2_3"] = "V2_3";
    Risc0Version["V3_0"] = "V3_0";
})(Risc0Version || (exports.Risc0Version = Risc0Version = {}));
var UltrahonkVariant;
(function (UltrahonkVariant) {
    UltrahonkVariant["Plain"] = "Plain";
    UltrahonkVariant["ZK"] = "ZK";
})(UltrahonkVariant || (exports.UltrahonkVariant = UltrahonkVariant = {}));
var Plonky2HashFunction;
(function (Plonky2HashFunction) {
    Plonky2HashFunction["Keccak"] = "Keccak";
    Plonky2HashFunction["Poseidon"] = "Poseidon";
})(Plonky2HashFunction || (exports.Plonky2HashFunction = Plonky2HashFunction = {}));
var Library;
(function (Library) {
    Library["arkworks"] = "arkworks";
    Library["gnark"] = "gnark";
    Library["snarkjs"] = "snarkjs";
})(Library || (exports.Library = Library = {}));
var CurveType;
(function (CurveType) {
    CurveType["bn128"] = "bn128";
    CurveType["bn254"] = "bn254";
    CurveType["bls12381"] = "bls12381";
})(CurveType || (exports.CurveType = CurveType = {}));
var AggregateSecurityRules;
(function (AggregateSecurityRules) {
    AggregateSecurityRules["Untrusted"] = "Untrusted";
    AggregateSecurityRules["OnlyOwner"] = "OnlyOwner";
    AggregateSecurityRules["OnlyOwnerUncompleted"] = "OnlyOwnerUncompleted";
})(AggregateSecurityRules || (exports.AggregateSecurityRules = AggregateSecurityRules = {}));
var ProofSecurityRules;
(function (ProofSecurityRules) {
    ProofSecurityRules["Untrusted"] = "Untrusted";
    ProofSecurityRules["OnlyOwner"] = "OnlyOwner";
    ProofSecurityRules["OnlyAllowlisted"] = "OnlyAllowlisted";
})(ProofSecurityRules || (exports.ProofSecurityRules = ProofSecurityRules = {}));
var Destination;
(function (Destination) {
    Destination["None"] = "None";
})(Destination || (exports.Destination = Destination = {}));
var RuntimeVersion;
(function (RuntimeVersion) {
    RuntimeVersion[RuntimeVersion["V1_3_0"] = 1003000] = "V1_3_0";
    RuntimeVersion[RuntimeVersion["V1_5_0"] = 1005000] = "V1_5_0";
})(RuntimeVersion || (exports.RuntimeVersion = RuntimeVersion = {}));
exports.PUBLIC_ZK_VERIFY_EVENTS = [
    ZkVerifyEvents.NewAggregationReceipt,
    ZkVerifyEvents.ProofVerified,
    ZkVerifyEvents.NewProof,
    ZkVerifyEvents.VkRegistered,
    ZkVerifyEvents.NewDomain,
    ZkVerifyEvents.DomainStateChanged,
    ZkVerifyEvents.AggregationComplete,
];
