// zkverifyjs/dist/commonjs/proofTypes/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TEEProcessor = exports.UltraHonkProcessor = exports.SP1Processor = exports.Plonky2Processor = exports.UltraPlonkProcessor = exports.Risc0Processor = exports.Groth16Processor = exports.FflonkProcessor = exports.EZKLProcessor = void 0;
var processor_1 = require("./ezkl/processor");
Object.defineProperty(exports, "EZKLProcessor", { enumerable: true, get: function () { return __importDefault(processor_1).default; } });
var processor_2 = require("./fflonk/processor");
Object.defineProperty(exports, "FflonkProcessor", { enumerable: true, get: function () { return __importDefault(processor_2).default; } });
var processor_3 = require("./groth16/processor");
Object.defineProperty(exports, "Groth16Processor", { enumerable: true, get: function () { return __importDefault(processor_3).default; } });
var processor_4 = require("./risc0/processor");
Object.defineProperty(exports, "Risc0Processor", { enumerable: true, get: function () { return __importDefault(processor_4).default; } });
var processor_5 = require("./ultraplonk/processor");
Object.defineProperty(exports, "UltraPlonkProcessor", { enumerable: true, get: function () { return __importDefault(processor_5).default; } });
var processor_6 = require("./plonky2/processor");
Object.defineProperty(exports, "Plonky2Processor", { enumerable: true, get: function () { return __importDefault(processor_6).default; } });
var processor_7 = require("./sp1/processor");
Object.defineProperty(exports, "SP1Processor", { enumerable: true, get: function () { return __importDefault(processor_7).default; } });
var processor_8 = require("./ultrahonk/processor");
Object.defineProperty(exports, "UltraHonkProcessor", { enumerable: true, get: function () { return __importDefault(processor_8).default; } });
var processor_9 = require("./tee/processor");
Object.defineProperty(exports, "TEEProcessor", { enumerable: true, get: function () { return __importDefault(processor_9).default; } });
// ADD_NEW_PROOF_TYPE
