// zkverifyjs/dist/esm/enums.js
export var TransactionType;
(function (TransactionType) {
  TransactionType[TransactionType["Verify"] = 1] = "Verify";
  TransactionType[TransactionType["VKRegistration"] = 2] = "VKRegistration";
  TransactionType[TransactionType["DomainRegistration"] = 3] = "DomainRegistration";
  TransactionType[TransactionType["DomainHold"] = 4] = "DomainHold";
  TransactionType[TransactionType["DomainUnregister"] = 5] = "DomainUnregister";
  TransactionType[TransactionType["Aggregate"] = 6] = "Aggregate";
  TransactionType[TransactionType["BatchVerify"] = 7] = "BatchVerify";
  TransactionType[TransactionType["DomainAddSubmitters"] = 8] = "DomainAddSubmitters";
  TransactionType[TransactionType["DomainRemoveSubmitters"] = 9] = "DomainRemoveSubmitters";
})(TransactionType || (TransactionType = {}));
export var TransactionStatus;
(function (TransactionStatus) {
  TransactionStatus["Broadcast"] = "broadcast";
  TransactionStatus["Dropped"] = "dropped";
  TransactionStatus["Error"] = "error";
  TransactionStatus["Finalized"] = "finalized";
  TransactionStatus["InBlock"] = "inBlock";
  TransactionStatus["Invalid"] = "invalid";
  TransactionStatus["Pending"] = "pending";
  TransactionStatus["Retracted"] = "retracted";
  TransactionStatus["Usurped"] = "usurped";
})(TransactionStatus || (TransactionStatus = {}));
export var ZkVerifyEvents;
(function (ZkVerifyEvents) {
  ZkVerifyEvents["AggregationComplete"] = "aggregationComplete";
  ZkVerifyEvents["NewAggregationReceipt"] = "newAggregationReceipt";
  ZkVerifyEvents["Broadcast"] = "broadcast";
  ZkVerifyEvents["CannotAggregate"] = "cannotAggregate";
  ZkVerifyEvents["DomainStateChanged"] = "domainStateChanged";
  ZkVerifyEvents["ErrorEvent"] = "error";
  ZkVerifyEvents["Finalized"] = "finalized";
  ZkVerifyEvents["IncludedInBlock"] = "includedInBlock";
  ZkVerifyEvents["NewDomain"] = "newDomain";
  ZkVerifyEvents["NewProof"] = "newProof";
  ZkVerifyEvents["ProofVerified"] = "proofVerified";
  ZkVerifyEvents["Unsubscribe"] = "unsubscribe";
  ZkVerifyEvents["VkRegistered"] = "vkRegistered";
})(ZkVerifyEvents || (ZkVerifyEvents = {}));
export var Risc0Version;
(function (Risc0Version) {
  Risc0Version["V2_1"] = "V2_1";
  Risc0Version["V2_2"] = "V2_2";
  Risc0Version["V2_3"] = "V2_3";
  Risc0Version["V3_0"] = "V3_0";
})(Risc0Version || (Risc0Version = {}));
export var UltrahonkVariant;
(function (UltrahonkVariant) {
  UltrahonkVariant["Plain"] = "Plain";
  UltrahonkVariant["ZK"] = "ZK";
})(UltrahonkVariant || (UltrahonkVariant = {}));
export var Plonky2HashFunction;
(function (Plonky2HashFunction) {
  Plonky2HashFunction["Keccak"] = "Keccak";
  Plonky2HashFunction["Poseidon"] = "Poseidon";
})(Plonky2HashFunction || (Plonky2HashFunction = {}));
export var Library;
(function (Library) {
  Library["arkworks"] = "arkworks";
  Library["gnark"] = "gnark";
  Library["snarkjs"] = "snarkjs";
})(Library || (Library = {}));
export var CurveType;
(function (CurveType) {
  CurveType["bn128"] = "bn128";
  CurveType["bn254"] = "bn254";
  CurveType["bls12381"] = "bls12381";
})(CurveType || (CurveType = {}));
export var AggregateSecurityRules;
(function (AggregateSecurityRules) {
  AggregateSecurityRules["Untrusted"] = "Untrusted";
  AggregateSecurityRules["OnlyOwner"] = "OnlyOwner";
  AggregateSecurityRules["OnlyOwnerUncompleted"] = "OnlyOwnerUncompleted";
})(AggregateSecurityRules || (AggregateSecurityRules = {}));
export var ProofSecurityRules;
(function (ProofSecurityRules) {
  ProofSecurityRules["Untrusted"] = "Untrusted";
  ProofSecurityRules["OnlyOwner"] = "OnlyOwner";
  ProofSecurityRules["OnlyAllowlisted"] = "OnlyAllowlisted";
})(ProofSecurityRules || (ProofSecurityRules = {}));
export var Destination;
(function (Destination) {
  Destination["None"] = "None";
})(Destination || (Destination = {}));
export var RuntimeVersion;
(function (RuntimeVersion) {
  RuntimeVersion[RuntimeVersion["V1_3_0"] = 1003000] = "V1_3_0";
  RuntimeVersion[RuntimeVersion["V1_5_0"] = 1005000] = "V1_5_0";
})(RuntimeVersion || (RuntimeVersion = {}));
export const PUBLIC_ZK_VERIFY_EVENTS = [ZkVerifyEvents.NewAggregationReceipt, ZkVerifyEvents.ProofVerified, ZkVerifyEvents.NewProof, ZkVerifyEvents.VkRegistered, ZkVerifyEvents.NewDomain, ZkVerifyEvents.DomainStateChanged, ZkVerifyEvents.AggregationComplete];
